pkg/controllers/machine_controller_volumes.go

```diff
@@ -83,75 +84,153 @@ func getLastVolumeSize(machine *api.Machine, volumeID string) *int64 {
 	return nil
 }
 
-func (r *MachineReconciler) attachDetachVolumes(ctx context.Context, log logr.Logger, machine *api.Machine, attacher VolumeAttacher) ([]api.VolumeStatus, error) {
+func fillStatusWithMountedVolumes(status map[string]*api.VolumeStatus, mounter VolumeMounter) error {
+	return mounter.ForEachVolume(func(volume *MountVolume) bool {
+		_, ok := status[volume.ComputeVolumeName]
+		if !ok {
+			volumeStatus := api.VolumeStatus{
+				Name:  volume.ComputeVolumeName,
+				State: api.VolumeStatePending,
+			}
+			status[volumeStatus.Name] = &volumeStatus
+		}
+		return true
+	})
+}
+
+// reconcileVolumes is doing attaching, detaching, deleting of volumes and it manage status of volumes
+func (r *MachineReconciler) reconcileVolumes(ctx context.Context, log logr.Logger, machine *api.Machine, attacher VolumeAttacher) ([]api.VolumeStatus, error) {
 	mounter := r.machineVolumeMounter(machine)
 	specVolumes := r.listDesiredVolumes(machine)
 
-	currentVolumeNames := sets.New[string]()
-	if err := attacher.ForEachVolume(func(volume *AttachVolume) bool {
-		currentVolumeNames.Insert(volume.Name)
-		return true
-	}); err != nil {
-		return nil, fmt.Errorf("error iterating attached volumes: %w", err)
+	currentStatus := machine.Status.GetVolumesAsMap()
+
+	attachedVolumes, err := attacher.ListVolumesAsMap()
+	if err != nil {
+		// missing list of attached volumes can affect deletion of volumes, so we cannot continue
+		return machine.Status.VolumeStatus, err
 	}
```

Comment:
I think it's a bad practice returning  "data" in an error case.

Reply:
here we can return nil, it isn't problem with new update status functionality.

Reply:
fixed